src/error.rs

```diff
@@ -59,32 +59,51 @@ impl_pyerr!(DimensionalityError);
 
 /// Represents that types of the given arrays do not match.
 #[derive(Debug)]
-pub struct TypeError {
-    from: String,
-    to: String,
+pub struct TypeError<'a> {
```

Comment:
Sadly, this addition of a lifetime parameter to `TypeError` means this cannot be part of a maintenance release because `TypeError` is part of the crate API even it is not part of any type signature as it is only ever produced wrapped within `PyErr`. :-(

Reply:
Decided to actually remove `DimensionalityError` and `TypeError` from the API. They are never used directly and cannot be recovered via downcasting either. So the additional API surface does not gain us anything.